pkg/policy/selectorcache.go

```diff
@@ -921,6 +978,28 @@ func (sc *SelectorCache) AddIdentitySelector(user CachedSelectionUser, selector
 	// is already cached, or is a new one).
 
 	sc.selectors[key] = newIDSel
+	sc.mutex.Unlock()
+
+	// wait for IDs to be allocated (if required).
+	// Ideally we defer this and let the caller decide when to wait, but the call stack
+	// is just too deep to ergonomically do this.
+	// So, wait for a short period of time for the ipcache to converge and, if
+	// necessary, give up and proceed. It's okay to proceed, since the selectorCache
+	// will *eventually* be updated and everything will converge.
+	start := time.Now()
+	if newIDSel.waitingForSelections > 0 {
+		// Give ourselves a 5-second timeout, just out of paranoia.
+		// If the ipcache is somehow not making progress, we don't want
+		// to block progress. This is safe, because the updated selections
+		// will be pushed back down in to the policy maps anyways
+		select {
+		case <-newIDSel.hasSelectionsCh:
+			log.WithField("prefixes", prefixes).Debug("Wait time for CIDR prefix allocation: ", time.Since(start))
+		case <-time.After(5 * time.Second):
+			log.WithField("prefixes", prefixes).Warn("Timed out waiting for CachedSelector to be populated! Policy may be temporarily inconsistent")
+		}
+	}
```

Comment:
I wonder if it's just much simpler to make a variant of `UpsertPrefixes` that's sychronous, now that the endpoint locks are no longer a concern in terms of deadlocks (`UpsertPrefixes` can take the endpoint lock for endpoint regeneration).

See https://github.com/christarazi/cilium/commit/df396ef4d545017acc1e852fc5f41671b63027db.

Reply:
Both of those proposed efforts are asynchronous -- the question is: who is doing the waiting. Is it the ipcache itself, or the consumer of the ipcache?

The advantage of moving the waiting in to the SelectorCache is that we can short-circuit waiting if we're already at the desired state. Say, for example, that the CIDR in question already exists. Then there's no need to wait for the ipcache to make progress; we can immediately continue. Likewise, if the ipcache makes partial progress but doesn't complete our unit of work, we could falsely assume that it is ready.

The disadvantage is that it is somewhat more complex; the waiting is now "far away" from the ipcache, code-wise. This means that debugging is more difficult, since the systems are loosely coupled.

Reply:
There's some obvious hand waving here and details to work out, but architecturally I think this is getting towards a more robust model that better separates and defines contracts between endpoints, identities, and policy. With a cleaner architecture, it will help avoid footguns like potentially locking up Cilium in case of errors and transient drops. 

My biggest concern the whole time with the selectorcache approach was the added complex the selectorcache layer which really had no business allocating new identities. It took a while before realizing what an alternative approach can look like, but going forward, I'm hoping the new model also lends itself a bit better to leverage `pkg/stream` objects.

Reply:
> That same logic from UpdatePolicyMaps() reaches through the selectorcache to this function here. When this line does close() of the channel, it asynchronously triggers the policymap sync, but does not guarantee that it completes before returning back up to the ipcache.

I don't think this is an issue.

Closing the channel means that the policy calculation is allowed to proceed. In other words, it now knows that the cached selector in question is up-to-date. This is actually irrelevant to policymap syncs, because we're not consuming deltas, we're generating the policy from scratch. In other words, the policy map for the newly-calculated policy will be correct.

The critical observation is that `ipcache.UpdatePolicyMaps` does two things:

1. Update the SelectorCache. This is `UpdateIdentities()`
2. Push deltas in to every `endpoint.desiredPolicy`. This is `endpointmanager.UpdatePolicyMaps()`

Step 1 will close the channel and allow policy distillation to proceed. But, again, I think this is safe because distillation only consumes the SelectorCache, not the bpf ipcache. In other words, while we have added a timing point, we haven't actually reordered any operations.

Reply:
Ah, so the reason why it's not an issue is because I assumed that `ipcache.UpdatePolicyMaps` went through the full policy calculation path which is part of endpoint regeneration, but it's actually the incremental path which directly updates the policymaps via `applyPolicyMapChanges()`.